diff --git a/cmd/httpserver/main.go b/cmd/httpserver/main.go
index f7fd738..ebf1708 100644
--- a/cmd/httpserver/main.go
+++ b/cmd/httpserver/main.go
@@ -20,7 +20,7 @@ func main() {
 		log.Fatalf("Error starting server: %v", err)
 	}
 	defer server.Close()
-	log.Println("Server started at address:", server.Addr().String())
+	log.Println("Server started at address:", server.Listener.Addr().String())
 
 	sigChan := make(chan os.Signal, 1)
 	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
@@ -41,7 +41,6 @@ func handler(res *response.Response, req *request.Request) {
 		res.Headers.Set("Content-Type", "text/plain")
 		res.Body("Whoopsie, my bad\n")
 	default:
-		res.SetStatusCode(response.StatusCode_200)
 		res.SetDefaultHeaders()
 		res.Headers.Set("Content-Type", "text/plain")
 		res.Body("All good frfr\n")
diff --git a/internal/response/response.go b/internal/response/response.go
index 22db654..cbd863f 100644
--- a/internal/response/response.go
+++ b/internal/response/response.go
@@ -57,7 +57,8 @@ type Response struct {
 
 func New() *Response {
 	return &Response{
-		Headers: headers.Create(),
+		Headers:    headers.Create(),
+		StatusCode: StatusCode_200,
 	}
 }
 
diff --git a/internal/server/server.go b/internal/server/server.go
index 3a1e620..684113e 100644
--- a/internal/server/server.go
+++ b/internal/server/server.go
@@ -5,14 +5,13 @@ import (
 	"log"
 	"net"
 	"sync/atomic"
-	"unsafe"
 
 	"boot.theprimeagen.tv/internal/request"
 	"boot.theprimeagen.tv/internal/response"
 )
 
 type Server struct {
-	Listener net.Listener //@ pointer?
+	Listener net.Listener
 	closed   atomic.Bool
 }
 
@@ -71,21 +70,14 @@ func (s *Server) Close() error {
 	return nil
 }
 
-func (s *Server) Addr() net.Addr {
-	return s.Listener.Addr()
-}
-
 func Serve(port int, handler Handler) (*Server, error) {
 	addr := fmt.Sprintf("127.0.0.1:%v", port)
 	l, err := net.Listen("tcp", addr)
 	if err != nil {
 		return nil, err
 	}
-	fmt.Printf("listener's type is %T\n", l)
-	fmt.Printf("listener's size is %v\n", unsafe.Sizeof(l))
 
-	s := &Server{Listener: l} //@ pointer?
-	// s.closed.Store(false)      // seems to be "false" by default
+	s := &Server{Listener: l}
 
 	go s.listen(handler)
 
