diff --git a/cmd/httpserver/main.go b/cmd/httpserver/main.go
index a408ec7..59ce99f 100644
--- a/cmd/httpserver/main.go
+++ b/cmd/httpserver/main.go
@@ -1,8 +1,10 @@
 package main
 
 import (
+	"boot.theprimeagen.tv/internal/request"
 	"boot.theprimeagen.tv/internal/response"
 	"boot.theprimeagen.tv/internal/server"
+	"io"
 	"log"
 	"os"
 	"os/signal"
@@ -15,18 +17,7 @@ import (
 const port = 42069
 
 func main() {
-	err := response.WriteStatusLine(os.Stdout, response.StatusCode_200)
-	if err != nil {
-		log.Print(err)
-	}
-
-	headers := response.GetDefaultHeaders(3)
-	err = response.WriteHeaders(os.Stdout, headers)
-	if err != nil {
-		log.Print(err)
-	}
-
-	server, err := server.Serve(port)
+	server, err := server.Serve(port, handler)
 	if err != nil {
 		log.Fatalf("Error starting server: %v", err)
 	}
@@ -37,6 +28,26 @@ func main() {
 	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
 	<-sigChan
 	log.Println("Server gracefully stopped")
+}
+
+func handler(w io.Writer, req *request.Request) *server.HandlerError {
+	var err *server.HandlerError = nil
+
+	switch req.RequestLine.RequestTarget {
+	case "/yourproblem":
+		err = &server.HandlerError{
+			Code:    response.StatusCode_400,
+			Message: "Your problem is not my problem\n",
+		}
+	case "/myproblem":
+		err = &server.HandlerError{
+			Code:    response.StatusCode_500,
+			Message: "Whoopsie, my bad\n",
+		}
+	default:
+		w.Write([]byte("All good frfr\n"))
+		err = nil
+	}
 
-	// time.Sleep(time.Second * 3)
+	return err
 }
diff --git a/internal/headers/headers.go b/internal/headers/headers.go
index e0d654a..ce00956 100644
--- a/internal/headers/headers.go
+++ b/internal/headers/headers.go
@@ -53,12 +53,11 @@ func (h *Headers) Parse(data []byte) (int, bool, error) {
 		}
 	}
 
-	headers := *h
 	fieldName = bytes.ToLower(fieldName)
-	if headers[string(fieldName)] == "" {
-		headers[string(fieldName)] = string(fieldValue)
+	if (*h)[string(fieldName)] == "" {
+		(*h)[string(fieldName)] = string(fieldValue)
 	} else {
-		headers[string(fieldName)] += ", " + string(fieldValue)
+		(*h)[string(fieldName)] += ", " + string(fieldValue)
 	}
 
 	return len(fieldLine) + 2, false, nil
diff --git a/internal/request/request.go b/internal/request/request.go
index dd9a6ad..6f829dc 100644
--- a/internal/request/request.go
+++ b/internal/request/request.go
@@ -3,7 +3,6 @@ package request
 import (
 	"bytes"
 	"errors"
-	"fmt"
 	"io"
 	"log"
 	"strconv"
@@ -50,24 +49,21 @@ func RequestFromReader(reader io.Reader) (*Request, error) {
 
 		n, err := reader.Read(buf[ri:])
 		if err != nil {
-			if err == io.EOF {
-				// Read returned EOF, which means that no more data is coming, but parser is expecting more...
-				return nil, fmt.Errorf("Incomplete data")
-			}
-			return nil, fmt.Errorf("Read: %v", err.Error())
+			return nil, err
 		}
 		ri += n
 
-		for req.status != reqStatusDone {
-			n, err = req.parse(buf[pi:ri])
-			if n == 0 {
-				if err != nil {
-					return nil, err
-				} else {
-					break
+		for {
+			bytesParsed, parseErr := req.parse(buf[pi:ri])
+			if bytesParsed == 0 {
+				if parseErr != nil {
+					return nil, parseErr
 				}
+
+				break
 			}
-			pi += n
+
+			pi += bytesParsed
 		}
 	}
 
@@ -116,6 +112,9 @@ func (r *Request) parse(d []byte) (int, error) {
 
 			r.status = reqStatusDone
 		}
+	case reqStatusDone:
+		n = 0
+		err = nil
 	default:
 		log.Panic("Unknown request status")
 	}
diff --git a/internal/response/response.go b/internal/response/response.go
index 4130995..8b6b76b 100644
--- a/internal/response/response.go
+++ b/internal/response/response.go
@@ -32,7 +32,7 @@ func (code StatusCode) String() string {
 
 func WriteStatusLine(w io.Writer, statusCode StatusCode) error {
 	statusLine := fmt.Sprintf("HTTP/1.1 %v %v\r\n", int(statusCode), statusCode)
-	_, err := w.Write([]byte(statusLine))
+	_, err := w.Write([]byte(statusLine)) //@ Does it write everything?
 	return err
 }
 
@@ -47,7 +47,7 @@ func GetDefaultHeaders(contentLen int) headers.Headers {
 func WriteHeaders(w io.Writer, headers headers.Headers) error {
 	for k, v := range headers {
 		header := fmt.Sprintf("%v: %v\r\n", k, v)
-		_, err := w.Write([]byte(header))
+		_, err := w.Write([]byte(header)) //@ Does it write everything?
 		if err != nil {
 			return err
 		}
diff --git a/internal/server/server.go b/internal/server/server.go
index 1498980..e33edce 100644
--- a/internal/server/server.go
+++ b/internal/server/server.go
@@ -1,8 +1,11 @@
 package server
 
 import (
+	"boot.theprimeagen.tv/internal/request"
 	"boot.theprimeagen.tv/internal/response"
+	"bytes"
 	"fmt"
+	"io"
 	"log"
 	"net"
 	"sync/atomic"
@@ -13,13 +16,20 @@ type Server struct {
 	closed   atomic.Bool
 }
 
-func (s *Server) listen() {
+type Handler func(w io.Writer, req *request.Request) *HandlerError
+
+type HandlerError struct {
+	Code    response.StatusCode
+	Message string
+}
+
+func (s *Server) listen(connHandler Handler) {
 	// It mostly doesn't run when the server is closed because this function is waiting on Accept() most of the time
 	//@ Is this fine?
-	// defer func() {
-	// 	(*s.listener).Close()
-	// 	log.Print("closed the listener")
-	// }()
+	defer func() {
+		(*s.listener).Close()
+		log.Print("closed the listener")
+	}()
 
 	for {
 		if s.closed.Load() {
@@ -34,20 +44,41 @@ func (s *Server) listen() {
 		if err != nil {
 			log.Fatal(err) //@ How to report this error?
 		}
+		fmt.Print("Accepted a connection.\n")
 
-		go s.handle(conn)
+		go s.handle(conn, connHandler)
 	}
 }
 
-func (s *Server) handle(conn net.Conn) {
-	body := "Oh yeah, baby!"
-	contentLen := len(body)
+func (s *Server) handle(conn net.Conn, connHandler Handler) {
+	defer func() {
+		conn.Close()
+		fmt.Print("Closed the connection.\n")
+	}()
 
-	err := response.WriteStatusLine(conn, response.StatusCode_500)
-	// err := response.WriteStatusLine(conn, response.StatusCode_200)
+	req, err := request.RequestFromReader(conn)
 	if err != nil {
-		log.Fatal(err) //@ How to report this error?
+		log.Print(err)
+		return
 	}
+	fmt.Printf("%v %v\n", req.RequestLine.Method, req.RequestLine.RequestTarget)
+	// for k, v := range req.Headers {
+	// 	fmt.Printf("%v: %v\n", k, v)
+	// }
+
+	var body bytes.Buffer
+	handlerErr := connHandler(&body, req)
+	if handlerErr != nil {
+		err = response.WriteStatusLine(conn, handlerErr.Code)
+		body.Write([]byte(handlerErr.Message))
+	} else {
+		err = response.WriteStatusLine(conn, response.StatusCode_200)
+	}
+	if err != nil {
+		panic("boom")
+	}
+
+	contentLen := body.Len()
 
 	headers := response.GetDefaultHeaders(contentLen)
 	err = response.WriteHeaders(conn, headers)
@@ -60,12 +91,10 @@ func (s *Server) handle(conn net.Conn) {
 		log.Fatal(err) //@ How to report this error?
 	}
 
-	_, err = conn.Write([]byte(body))
+	_, err = body.WriteTo(conn)
 	if err != nil {
 		log.Fatal(err) //@ How to report this error?
 	}
-
-	conn.Close()
 }
 
 func (s *Server) Close() error {
@@ -78,14 +107,17 @@ func (s *Server) Addr() net.Addr {
 	return (*s.listener).Addr()
 }
 
-func Serve(port int) (*Server, error) {
+func Serve(port int, connHandler Handler) (*Server, error) {
 	addr := fmt.Sprintf("127.0.0.1:%v", port)
 	l, err := net.Listen("tcp", addr)
 	if err != nil {
 		return nil, err
 	}
+
 	s := &Server{listener: &l} //@ pointer?
-	s.closed.Store(false)      // seems to be "false" by default
-	go s.listen()
+	// s.closed.Store(false)      // seems to be "false" by default
+
+	go s.listen(connHandler)
+
 	return s, nil
 }
